Kubernetes provider.

mateusz-blaszkowski opened the pull request:
Instead of VMs in Kubernetes we have a concept of PODs. Create() method
creates a POD with a single container inside. Access to this container is
done through the Service. Authentication is done via Secret - SSH key is
base64 encoded and stored in a file which is later mounted as a volume.

Additionaly, Ceph support was added. You can specify IP addresses of Ceph
Monitors. When scratch disk is needed, RBD image will be created with ext4
filesystem. This volume has to be created _before_ the actual POD launch.
Ceph volume is currently the only option if the benchmark requires scratch
disk but I plan to add local disk support in the next check-ins.

Signed-off-by: Mateusz Blaszkowski <mateusz.blaszkowski@intel.com>

mateusz-blaszkowski commented:
Hi

I am covered by the 'Google Corporate CLA	Intel CLA Submissions	Nov 25, 2014 14:48 PST' corporate agreement. My email in 'Signed-off by' is also correct. Can you help me with CLA?

voellm commented:
We'll pick this up pretty soon.  Thanks for working on it.

cmccoy commented on perfkitbenchmarker/kubernetes/kubernetes_virtual_machine.py:
```diff
@@ -0,0 +1,482 @@
+                  'create', '-f', '-']
+    process = Popen(create_cmd, stdout=PIPE, stdin=PIPE, stderr=STDOUT)
+    create_rc_body = self._BuildPodBody()
+    output = process.communicate(input=create_rc_body)
```
What happens if the command fails? Could you check the exit status here?

cmccoy commented on perfkitbenchmarker/kubernetes/kubernetes_virtual_machine.py:
```diff
@@ -0,0 +1,482 @@
+    Creates a Service (POD accessor).
+    """
+    create_cmd = [FLAGS.kubectl, '--kubeconfig=%s' % FLAGS.kubeconfig,
+                  'create', '-f', '-']
```
As above

cmccoy commented on perfkitbenchmarker/kubernetes/kubernetes_virtual_machine.py:
```diff
@@ -0,0 +1,482 @@
+            and pod_status == "Running"):
+          logging.info("POD is up and running.")
+          return
+    raise Exception()
```
Could you add an error message here?

cmccoy commented on perfkitbenchmarker/kubernetes/kubernetes_virtual_machine.py:
```diff
@@ -0,0 +1,482 @@
+        logging.info("Endpoint found. Service is successfully matched with "
+                     "POD.")
+        return
+    raise Exception()
```
As above

cmccoy commented:
@mateusz-blaszkowski - thanks for this! Looks great. Do you think you could add some more logging around the commands to `kubectl`? I'm concerned that errors will be difficult to track down, since process stdout and stderr is captured, and some exit codes aren't checked.

mateusz-blaszkowski commented:
@cmccoy - done. Although still only one storage type is supported right now, the code is ready to be merged. In the near future I will be working on adding more storage options (Kubernetes Persistent Volumes first).

cmccoy commented on README.md:
```diff
@@ -159,6 +159,57 @@ echo topsecretpassword > ~/.config/openstack-password.txt
 
+## Kubernetes configuration and credentials
+Perfkit uses `kubectl` binary in order to communicate with Kubernetes cluster - you need to pass the path to `kubectl` binary using `--kubectl` flag.  
+Authentication to Kubernetes cluster is done via `kubeconfig` file. Its path is passed using `--kubeconfig` flag.
```
Nitpick (not necessary for merge) - could you link to the kubeconfig documentation here? https://github.com/kubernetes/kubernetes/blob/release-1.0/docs/user-guide/kubeconfig-file.md

cmccoy commented:
@mateusz-blaszkowski: sorry for the delay here - I'm working on trying this out.

mateusz-blaszkowski commented:
@cmccoy - sure, I have updated the documentation and also added a brief description of creating a Ceph Secret. If you need any help with running the benchmarks - don't hesitate to ask :) We run benchmarks on CoreOS + Ceph cluster. Currently only Ceph is supported so if a test requires scratch volume and you don't have K8S integrated with Ceph, the test will not be executed.

voellm commented:
I plan to roll the next release today.  Is this ready to integrate?

mateusz-blaszkowski commented:
Yes, it's ready.

cmccoy commented:
I haven't had the chance to test this out yet, sorry. The code looks good - two comments (maybe for the next PR):

* Could `--kubectl` default to `kubectl` (assuming it's on the user's PATH)?
* Could `--kubeconfig` have a reasonable default? e.g., `~/.kube/config`?

@mateusz-blaszkowski: Have you tested the latest revision? Are any benchmarks unsupported?
Otherwise, LGTM.

mateusz-blaszkowski commented:
@cmccoy - I haven't executed all the tests yet. This is the list of tests that passed without any problems so far:
aerospike, block_storage_workload, cassandra_ycsb, cluster_boot, coremark, fio, iperf, oldisim, ping, redis, unixbench

Special cases:
bonnie++ - currently not supported in container
cassandra_stress - I've got an exception: `Error: GetScratchDir(disk_num=0) is invalid, max disk_num is 0` . Looks like a bug in a benchmark. Cassandra_stress shouldn't require scratch disks, however GetScratchDir() is called
cloud_bigtable_ycsb - not executed because requires preconfigured Bigtable cluster
copy_throughput - gives invalid results when running in non-privileged mode
hadoop_terasort - hangs on `python /tmp/pkb/wait_for_command.py` command, not sure what's going on
speccpu2006 - not executed because it's not available for free

I haven't executed the rest of the bechmarks - it's in progress. Tomorrow I will rebase to the latest code and I will let you know what's the status.

cmccoy commented:
That sounds great. Happy to merge whenever you give the word.

Regarding Cassandra stress - the scratch disk requirement is correct - the benchmark creates both a Cassandra cluster and client nodes to stress the cluster.

Thanks for this and for testing!

mateusz-blaszkowski commented:
Are you sure? I haven't investigated it so deeply, but I can see the following in cassandra_stress benchmark:

BENCHMARK_INFO = {'name': 'cassandra_stress',
                  'description': 'Benchmark Cassandra using cassandra-stress',
                  **'scratch_disk': False,**
                  'num_machines': DEFAULT_CLUSTER_SIZE}

I have tested with the latest revision and it looks good. I verified a couple of new benchmarks:
* sysbench_oltp - pass
* scimark2 - pass
* mysql_service - Kubernetes not supported (it looks like this benchmark requires a special support for any new provider - check CheckPrerequisites() method in mysql_service_benchmark.py)
* silo - failed with `Clone of 'git://github.com/kohler/masstree-beta.git' into submodule path 'masstree' failed`. I run the test behind the proxy and this is the case. I would have changed the path to Git repository to https:// but it is hidden somewhere in 'dbtest' (look a the command which failed: `cd /tmp/pkb/silo && MODE=perf DEBUG=0 CHECK_INVARIANTS=0 make          -j80 dbtest`). Oh, i found that the exact path is specified here: https://github.com/stephentu/silo/blob/cc11ca1ea949ef266ee12a9b1c310392519d9e3b/.gitmodules
* object_storage_service - not executed because requires GCP / AZURE / AWS storage provider
* hbase - failed. Tries to download package from http://apache.mirrors.tds.net/hbase/1.0.1.1/hbase-1.0.1.1-bin.tar.gz but it doesn't exist any more. **This should be filed as an issue** :)

Apart from these, I have verified some benchmarks which previously passed - they also pass with the latest revision. IMO ready for merge!

cmccoy commented:
Thanks for bringing up the issue with Cassandra - I opened #490 to track that.

Thank you so much, @mateusz-blaszkowski - this is a great addition to have.

voellm commented:
@connormccoy Please open issues for things that should be fixed.  For
example moving the silo link to https, refactoring the mysql_service so it
does not need updating for every provider, etc.  Thanks!

On Fri, Sep 25, 2015 at 7:51 AM, Connor McCoy <notifications@github.com>
wrote:

> Merged #475
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/475>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/475#event-419239213>
> .
>



-- 
Anthony F. Voellm (aka Tony)
Google Voice:  (650) 516-7382
<https://www.google.com/voice/b/0?pli=1#phones>
Blog: http://perfguy.blogspot.com/
Benchmark like a pro... PerfKitBenchmarker
<https://github.com/GoogleCloudPlatform/PerfKitBenchmarker>

cmccoy commented:
@voellm:

* Opened #491 to track silo benchmark changes.
* I'm sending a PR for the HBase issue now.
* I don't think that `cloud_bigtable` or `mysql_service` need updating for Kubernetes. Both test managed services only available on some cloud providers. `sysbench_oltp` is equivalent for running self-managed MySQL.  Cloud Bigtable *could* be updated to use a provided service account key, but it would be testing a remote service, rather than the Kubernetes cluster it's running on.

voellm commented:
The impact for Cloud Bigtable would be the networking path of which
Kubernetes is in the path... so it does have some value but not as much as
pure networking micro benchmarks.   So at best a P2.

On Fri, Sep 25, 2015 at 8:21 AM, Connor McCoy <notifications@github.com>
wrote:

> @voellm <https://github.com/voellm>:
>
>    - Opened #491
>    <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/491>
>    to track silo benchmark changes.
>    - I'm sending a PR for the HBase issue now.
>    - I don't think that cloud_bigtable or mysql_service need updating for
>    Kubernetes. Both test managed services only available on some cloud
>    providers. sysbench_oltp is equivalent for running self-managed MySQL.
>    Cloud Bigtable *could* be updated to use a provided service account
>    key, but it would be testing a remote service, rather than the Kubernetes
>    cluster it's running on.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/475#issuecomment-143250633>
> .
>



-- 
Anthony F. Voellm (aka Tony)
Google Voice:  (650) 516-7382
<https://www.google.com/voice/b/0?pli=1#phones>
Blog: http://perfguy.blogspot.com/
Benchmark like a pro... PerfKitBenchmarker
<https://github.com/GoogleCloudPlatform/PerfKitBenchmarker>
